---
title: "Allele-specific ChiPseq analysis. Check in on Genotype data"
author: "Anastasiia Hryhorzhevska"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{r, warning=FALSE, message=FALSE, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
  BiocManager::install(version = "3.12")
}
# library(ChIPseeker) 
library(DiffBind)  
library(GenomicFeatures)
library(GenomicRanges)
library(dplyr)
```

```{r, include=FALSE}
wd <- "/Users/anastasiia_hry/bio/code/mpip/signe_chip-seq/"
setwd(wd)

result.dir    <- paste0(wd, "03_results/")
src.data.dir  <- paste0(wd, "01_source_data/") 
starr.data.fn <- paste0(src.data.dir, "01_allele_dependent_STARR.txt") 
grsnp.data.fn <- paste0(src.data.dir, "02_GR_SNPs_hg19.txt") 
```

# 1. Load data

```{r, include=FALSE}
peaks.list <- readRDS(paste0(result.dir, "GRanges_overkal_0_ChIPseq_obj_Signe.rds"))

starr.df <- read.csv(starr.data.fn, sep = "\t")
grsnp.df <- read.csv(grsnp.data.fn, sep = "\t")
```

# 2. Convert STARR and GR SNP data into GRange object

## 2.1. STARR data

```{r}
starr.df$chr <- substr(starr.df$chr, 4, length(starr.df$chr))

starr <- GRanges(
  starr.df$chr,
  IRanges(start = starr.df$position, end = starr.df$position),
  strand = "*"
)


starr <- makeGRangesFromDataFrame(starr.df, start.field = "position", end.field = "position", seqnames.field = c("chr"))
names(starr) <- starr.df$name
starr.hg38 <- deepblue_liftover(starr, source = "hg19", target = "hg38", collapse = T)
# seqnames(starr.hg38) <- substr(seqnames(starr.hg38), 4, length(seqnames(starr.hg38)))
seqlevels(starr.hg38) <- substr(seqlevels(starr.hg38), 4, length(seqlevels(starr.hg38)))
starr.hg38
```

## 2.2.GR-SNP data

```{r}
grsnp.df$CHROM <- paste0("chr", grsnp.df$CHROM)
grsnp <- makeGRangesFromDataFrame(grsnp.df, start.field = "POS", end.field = "POS", seqnames.field = c("CHROM"))
names(grsnp) <- grsnp.df$ID
grsnp.hg38 <- deepblue_liftover(grsnp , source = "hg19", target = "hg38", collapse = T)
# seqnames(starr.hg38) <- substr(seqnames(starr.hg38), 4, length(seqnames(starr.hg38)))
seqlevels(grsnp.hg38) <- substr(seqlevels(grsnp.hg38), 4, length(seqlevels(grsnp.hg38)))
starr.hg38

```

# 3. Overlap STARR data with peaksets

## 3.1. Check if SNPs are in the binding dits DEX

##### For veh or veh-dex peaksets, there is no SNPs in the binding sites

```{r}
r1 <- starr.hg38
r2 <- peaks.list$dex
overlap <- findOverlaps(r1, r2, select = "all")
overlap
```

Make vector of SNPs to peaks

```{r}
hits <- names(r2)[subjectHits(overlap)]
names(hits) <- names(r1)[queryHits(overlap)]
hits
```

```{r}
r1[names(hits),]
```

```{r}
r2[hits, ]
```

## Find the distances to the nearest SNP for each peak for DEX

```{r}
nearest(r1, r2)[1:20]
```

```{r}
dist.to.nearest.obj <- distanceToNearest(r1, r2)
dist.to.nearest.obj
```

```{r}
dists <- dist.to.nearest.obj@elementMetadata@listData$distance
summary(dists)
```

```{r}
# hist(dists,xlab = "dist to nearest peak", main = "Distances")
sort(dists)[1:20]
```

```{r}
hits <- names(r2)[subjectHits(dist.to.nearest.obj)]
names(hits) <- names(r1)[queryHits(dist.to.nearest.obj)]
hits[1:3 ]
```

```{r}
r1[names(hits)[1], ]
```

```{r}
r2[hits[1], ]
```

Combine SNP, sites and distances into one dataframe

```{r}
# dist.df <- cbind(as.data.frame(hits), dists)
# dist.df

starr.dist <- cbind(as.data.frame(r1[names(hits), ]), dists) %>% 
 mutate(CHR = seqnames, POS = start, DIST = dists) %>% 
 select(CHR, POS, DIST) 

starr.bs.df <- cbind(starr.dist, as.data.frame(r2[hits, ]@seqnames), as.data.frame(r2[hits, ]@ranges))
colnames(starr.bs.df) <- c("CHR", "POS", "DIST", "PEAK_CHR", "PEAK_START", "PEAK_END", "PEAK_WIDTH", "PEAK_NAME")
starr.bs.df["TREATMENT"] <- "DEX"
starr.bs.df <- starr.bs.df[order(starr.bs.df$DIST), ]
starr.bs.df

# write.csv2(starr.bs.df, paste0(result.dir, "01_STARR_SNP_TO_ChIP_BINDING_SITES_DISTANCES_DEX.csv"), row.names = T)
```

## Find the distances to the nearest SNP for each peak for VEH

```{r}
r1 <- starr
r2 <- peaks.list$veh
dist.to.nearest.obj <- distanceToNearest(r1, r2)
dists <- dist.to.nearest.obj@elementMetadata@listData$distance
summary(dist.to.nearest.obj)
```

```{r}
hits <- names(r2)[subjectHits(dist.to.nearest.obj)]
names(hits) <- names(r1)[queryHits(dist.to.nearest.obj)]
```

Combine SNP, sites and distances into one dataframe

```{r}
# dist.df <- cbind(as.data.frame(hits), dists)
# dist.df

starr.dist <- cbind(as.data.frame(r1[names(hits), ]), dists) %>% 
  dplyr::mutate(CHR = seqnames, POS = start, DIST = dists) %>% 
  dplyr::select(CHR, POS, DIST) 

starr.bs.df <- cbind(starr.dist, as.data.frame(r2[hits, ]@seqnames), as.data.frame(r2[hits, ]@ranges))
colnames(starr.bs.df) <- c("CHR", "POS", "DIST", "PEAK_CHR", "PEAK_START", "PEAK_END", "PEAK_WIDTH", "PEAK_NAME")
starr.bs.df["TREATMENT"] <- "VEH"
starr.bs.df <- starr.bs.df[order(starr.bs.df$DIST), ]
starr.bs.df

 # write.csv2(starr.bs.df, paste0(result.dir, "02_STARR_SNP_TO_ChIP_BINDING_SITES_DISTANCES_VEH.csv"), row.names = T)
```

## Find the distances to the nearest SNP for each peak for VEH

```{r}
r1 <- starr
r2 <- peaks.list$`veh-dex`
dist.to.nearest.obj <- distanceToNearest(r1, r2)
dists <- dist.to.nearest.obj@elementMetadata@listData$distance
summary(dist.to.nearest.obj)
```

```{r}
hits <- names(r2)[subjectHits(dist.to.nearest.obj)]
names(hits) <- names(r1)[queryHits(dist.to.nearest.obj)]
```

Combine SNP, sites and distances into one dataframe

```{r}
# dist.df <- cbind(as.data.frame(hits), dists)
# dist.df

starr.dist <- cbind(as.data.frame(r1[names(hits), ]), dists) %>% 
  dplyr::mutate(CHR = seqnames, POS = start, DIST = dists) %>% 
  dplyr::select(CHR, POS, DIST) 

starr.bs.df <- cbind(starr.dist, as.data.frame(r2[hits, ]@seqnames), as.data.frame(r2[hits, ]@ranges))
colnames(starr.bs.df) <- c("CHR", "POS", "DIST", "PEAK_CHR", "PEAK_START", "PEAK_END", "PEAK_WIDTH", "PEAK_NAME")
starr.bs.df["TREATMENT"] <- "VEH-DEX"
starr.bs.df <- starr.bs.df[order(starr.bs.df$DIST), ]
starr.bs.df

# write.csv2(starr.bs.df, paste0(result.dir, "03_STARR_SNP_TO_ChIP_BINDING_SITES_DISTANCES_VEH-DEX.csv"), row.names = T)
```

# 4. Overlap all GR-SNPs data with peaksets

```{r}
r1 <- grsnp
r2 <- peaks.list$dex
overlap <- findOverlaps(r1, r2, select = "all")
overlap
```

Make vector of SNPs to peaks

```{r}
hits <- names(r2)[subjectHits(overlap)]
names(hits) <- names(r1)[queryHits(overlap)]
hits
```

```{r}
r1[names(hits),]
```

```{r}
r2[hits, ]
```

## Find the distances to the nearest SNP for each peak for DEX

```{r}
r1 <- grsnp
r2 <- peaks.list$dex
dist.to.nearest.obj <- distanceToNearest(r1, r2)
dists <- dist.to.nearest.obj@elementMetadata@listData$distance
summary(dist.to.nearest.obj)
```

```{r}
hits <- names(r2)[subjectHits(dist.to.nearest.obj)]
names(hits) <- names(r1)[queryHits(dist.to.nearest.obj)]
hits[1:10]
```

Combine SNP, sites and distances into one dataframe

```{r}
# dist.df <- cbind(as.data.frame(hits), dists)
# dist.df

grsnp.dist <- cbind(as.data.frame(r1[names(hits), ]), dists) %>% 
  dplyr::mutate(CHR = seqnames, POS = start, DIST = dists) %>% 
  dplyr::select(CHR, POS, DIST) 

grsnp.bs.df <- cbind(grsnp.dist, as.data.frame(r2[hits, ]@seqnames), as.data.frame(r2[hits, ]@ranges))
colnames(grsnp.bs.df) <- c("CHR", "POS", "DIST", "PEAK_CHR", "PEAK_START", "PEAK_END", "PEAK_WIDTH", "PEAK_NAME")
grsnp.bs.df["TREATMENT"] <- "DEX"
grsnp.bs.df <- grsnp.bs.df[order(grsnp.bs.df$DIST), ]
grsnp.bs.df

# write.csv2(grsnp.bs.df, paste0(result.dir, "04_GRSNP_SNP_TO_ChIP_BINDING_SITES_DISTANCES_DEX.csv"), row.names = T)
```

## Find the distances to the nearest SNP for each peak for VEH

```{r}
r1 <- grsnp
r2 <- peaks.list$veh
dist.to.nearest.obj <- distanceToNearest(r1, r2)
dists <- dist.to.nearest.obj@elementMetadata@listData$distance
summary(dist.to.nearest.obj)
```

```{r}
hits <- names(r2)[subjectHits(dist.to.nearest.obj)]
names(hits) <- names(r1)[queryHits(dist.to.nearest.obj)]
hits[1:10]
```

Combine SNP, sites and distances into one dataframe

```{r}
# dist.df <- cbind(as.data.frame(hits), dists)
# dist.df

grsnp.dist <- cbind(as.data.frame(r1[names(hits), ]), dists) %>% 
  dplyr::mutate(CHR = seqnames, POS = start, DIST = dists) %>% 
  dplyr::select(CHR, POS, DIST) 

grsnp.bs.df <- cbind(grsnp.dist, as.data.frame(r2[hits, ]@seqnames), as.data.frame(r2[hits, ]@ranges))
colnames(grsnp.bs.df) <- c("CHR", "POS", "DIST", "PEAK_CHR", "PEAK_START", "PEAK_END", "PEAK_WIDTH", "PEAK_NAME")
grsnp.bs.df["TREATMENT"] <- "VEH"
grsnp.bs.df <- grsnp.bs.df[order(grsnp.bs.df$DIST), ]
grsnp.bs.df

write.csv2(grsnp.bs.df, paste0(result.dir, "05_GRSNP_SNP_TO_ChIP_BINDING_SITES_DISTANCES_VEH.csv"), row.names = T)
```

## Find the distances to the nearest SNP for each peak for DEX-VEH

```{r}
r1 <- grsnp
r2 <- peaks.list$`veh-dex`
dist.to.nearest.obj <- distanceToNearest(r1, r2)
dists <- dist.to.nearest.obj@elementMetadata@listData$distance
summary(dist.to.nearest.obj)
```

```{r}
hits <- names(r2)[subjectHits(dist.to.nearest.obj)]
names(hits) <- names(r1)[queryHits(dist.to.nearest.obj)]
hits[1:10]
```

Combine SNP, sites and distances into one dataframe

```{r}
# dist.df <- cbind(as.data.frame(hits), dists)
# dist.df

grsnp.dist <- cbind(as.data.frame(r1[names(hits), ]), dists) %>% 
  dplyr::mutate(CHR = seqnames, POS = start, DIST = dists) %>% 
  dplyr::select(CHR, POS, DIST) 

grsnp.bs.df <- cbind(grsnp.dist, as.data.frame(r2[hits, ]@seqnames), as.data.frame(r2[hits, ]@ranges))
colnames(grsnp.bs.df) <- c("CHR", "POS", "DIST", "PEAK_CHR", "PEAK_START", "PEAK_END", "PEAK_WIDTH", "PEAK_NAME")
grsnp.bs.df["TREATMENT"] <- "VEH-DEX"
grsnp.bs.df <- grsnp.bs.df[order(grsnp.bs.df$DIST), ]
grsnp.bs.df

write.csv2(grsnp.bs.df, paste0(result.dir, "06_GRSNP_SNP_TO_ChIP_BINDING_SITES_DISTANCES_VEH_DEX.csv"), row.names = T)
```